### ISSUE TITLE
shell: dropdown cannot escape out of the SSH keys dialog

### ISSUE THREAD

**Issue by KKoukiou** [cockpit-project/cockpit#15769]
Cockpit version: 243
Page: Shell

![Screen Shot 2021-04-31 at 14 12 50](https://user-images.githubusercontent.com/14921356/116693537-5e8b4d00-a9be-11eb-9724-31522f62be69.png)

**Comment by KKoukiou** [comment]
This does not have an easy fix, it's some old dropdown based on boostrap and we did not really find a way to make it escape the PF4 modal. 
Let's move pkg/shell/credentials.js to React at some point.

**Comment by garrett** [comment]
#15912 had me looking at this dialog again and I saw this same issue.

If there isn't a good PF fix, then what I could suggest is that we turn off overflow for the dialog's content and turn on overflow auto for the list with a max-height. This would allow the popup to escape the boundaries and still let someone scroll a long list of ssh keys.

This dialog probably needs to be revised anyway, but this workaround fix should be easy enough that perhaps I should just do this and send a PR. Right now, it's otherwise unusable.

**Comment by KKoukiou** [comment]
@garrett if you have a workaround please git grep for dropdown-menu bootstrap class in our repo - and apply this workaround in all dropdowns in new modals.

**Comment by garrett** [comment]
There's no generic solution here except to use HTML forms, which doesn't work in this case.

The only actual solution still is:

1. Use HTML form `<select>`

That's it. That's the _only_ solution available.

However, workarounds include:

1. Use HTML form instead of the widget you actually want. (Meaning: If you want or need another element, you can't use it.)
2. Detatch the popup part and move it outside of the scrollable region. (This is a _horrible_ hack).
   - There's no association between the widget and its dropdown
   - Fixing the interaction of all sorts of interaction (screenreaders, keyboard usage, mouse actions, etc.) is complicated
      - Focus has to be tracked and manipulated
      - Position has to be calculated and updated on initial draw and also using scroll and resize event handling
3. Don't have the region where the popup be scrollable; if something needs to scroll, then have just that scroll instead (or a container for it)... but this is custom in every case; there's nothing generic that can be done.

So, we have a choice; we can do one of the following:

1. Keep ignoring the issue, letting our users suffer (not a solution)
2. Rethink the UI to make it not depend on a non-HTML dropdown
3. Plead with the PatternFly folks to hook up a horrible hack that covers all cases and wait for years for it to be implemented, if ever
4. Implement case-by-case custom workaround (# 3 from the workaround list above)

There's nothing else that can be done. This is why we have had issues like this, and this is why they stagnate. We've been doing # 1 and # 3, but that's not good enough.

**Comment by garrett** [comment]
Note: This problem can show up anywhere there's an element that wants to escape a scrollable region. It's most likely in smaller scrollable regions, like modals (which we use a lot, so we're probably more prone to see it than a lot of other projects using PF who might not use as many modals).

**Comment by garrett** [comment]
Wow, this dialog is such a mess of old code. I'm just trying to correct the overflow, but I'm testing it in different sizes and that's making me see how broken it is in mobile, which is causing me to edit the CSS.

I don't think I can properly fix this unless it's rewritten, and then I'd be happy to. As-is, I'm making way too many changes to legacy code and "shaving the yak" to make this overflow work as expected.

In other words: The overflow fix itself is simple, but it's "unraveling the whole sweater" by tugging on this one thread.

I'm happy to help with a modern re-implementation though!

**Comment by KKoukiou** [comment]
@garrett It's indeed old - inconsistent with the new PF4 widgets and I would be really suprised if back then when it was implemented it was even tested on mobile. +1 for rewrite.

**Comment by garrett** [comment]
Meanwhile, I stripped away everything but the basics just to fix this issue and will send up a PR.

It's basically just an emergency triage of UI that's 100% broken everywhere, to make it work fine in desktop and be a little less broken in smaller sizes.
